Draft for section 7.4 on transition of 2.0 and 2.1 in distribution

oxisto opened the pull request:
Fixes #843: Adding new draft for chapter on 2.0/2.1 transition in distribution

TODOs:

- [x] Check for consistency, spellcheck, style, SHOULD/SHOULD NOT, etc.
- [x] Reference CSAF2.0/2.1 converter conformance targets
- [x] Add `maintained_until` and `maintained_from` fields
- [x] Adding requirements for the above fields
- [x] Add examples (not link them in the text yet)

tschmidtb51 commented:
We'll keep that in draft - so you can rebase it after #990 is merged. (No rule without exception ;-))

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,25 @@
+  The content of the resource MUST be equal to the resource accessible at `/.well-known/csaf/v2.0/provider-metadata.json`.
+  For file-based distribution servers, this can for example be achieved by a symlink.
+- After the transition period has ended, the contents of the `provider-metadata.json` at `/.well-known/csaf/provider-metadata.json` SHALL be replaced 
+  with the contents of the resource available at `/.well-known/csaf/v2.1/provider-metadata.json`.
```
> Would we want to get rid of the `/.well-known/csaf/v2.1/provider-metadata.json` so that we only have `/.well-known/csaf/provider-metadata.json` at the end?

Honestly, I don't care. I would argue to keep it in that directory as the next transition is coming at some point ;-)

> What would be its canonical URL? Back to our version-less well-known

In that case: yes.

> Would it actually make sense to discover the version specific well-knowns, even if they are not in the security.txt?

I would argue no (at least not for now). If you have it and do not announce it in the security.txt - there is probably a reason for that. So, by choosing what I announce in the security.txt I have control over what information is distributed.

oxisto commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,25 @@
+  Its `canonical_url` must be set correctly and the property `maintained_until` SHOULD be set to an appropriate date in the future.
+  The url MUST also be added as an entry to the security.txt, if used.
+- A `provider-metadata.json` in conformance to CSAF 2.1 SHOULD be placed at `/.well-known/csaf/v2.1/provider-metadata.json`.
+  Its `canonical_url` must be set correctly, optionally the property `maintained_from` can be set to an appropriate date in the future, 
```
I wonder if we actually also should specify a `maintained_until` that ends with the transition period because after the transition period we don't necessarily want users to access it with the 2.1-WKURL but rather with the versionless-WKURL. In this case we might not need `maintained_from` at all. And probably `available_until` might be a better property name?

tschmidtb51 reviewed (changes requested):
Some tweaks are still to be done

oxisto commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,27 @@
+    - Optionally the property `maintained_from` can be set to an appropriate date in the future, if the service is not considered stable yet.
+    - The url SHALL also be added to the security.txt, if used.
+- At the begin of the transition period, a `provider-metadata.json` in conformance to CSAF 2.0 SHOULD be placed at `/.well-known/csaf/provider-metadata.json`.
+  - The content of the resource SHALL be equal to the resource accessible at `/.well-known/csaf/v2.0/provider-metadata.json`.
+  - For file-based distribution servers, this can for example be achieved by a symlink.
```
comment in wrong line?

oxisto commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,27 @@
+- At the begin of the transition period, a `provider-metadata.json` in conformance to CSAF 2.0 SHOULD be placed at `/.well-known/csaf/provider-metadata.json`.
+  - The content of the resource SHALL be equal to the resource accessible at `/.well-known/csaf/v2.0/provider-metadata.json`.
+  - For file-based distribution servers, this can for example be achieved by a symlink.
+- After the transition period has ended, the contents of the `provider-metadata.json` at `/.well-known/csaf/provider-metadata.json` SHALL be replaced
+  with the contents of the resource available at `/.well-known/csaf/v2.1/provider-metadata.json`.
```
this sort of leaves open how this is communicated I fear. Why is the transition period not over at this point?

sthagen reviewed (approved):
LGTM. Stefan will fix the mappings and sync the user facing delivery items after merge in the editorial 2025-05-28 branch.

tschmidtb51 commented:
- [ ] Please add the following text to `distributing-01-requirments.md` line 167:

```

As specified in [sec](#transition-between-csaf-20-and-csaf-21), the value of `canonical_url` MAY differ from the URL that was requested as a part of this requirement.
Such state is intended and MUST NOT be reported as error.

```

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,28 @@
+  - For file-based distribution servers, this MAY for example be achieved by a symlink.
+- After the transition period has ended, the contents of the `provider-metadata.json` at `/.well-known/csaf/provider-metadata.json` SHALL be
+  replaced with the contents of the resource available at `/.well-known/csaf/v2.1/provider-metadata.json`.
+  - The canonical URL SHALL be set to the URL corresponding to `/.well-known/csaf/provider-metadata.json`.
```
Either we remove that sentence or convert it into a MAY statement.

oxisto replied:
Which sentence are you referring to?

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,28 @@
+  - The URL SHALL also be added to the security.txt, if used.
+- At the begin of the transition period, a `provider-metadata.json` in conformance to CSAF 2.0 SHOULD be placed at `/.well-known/csaf/provider-metadata.json`.
+  - The content of the resource SHALL be equal to the resource accessible at `/.well-known/csaf/v2.0/provider-metadata.json`.
+  - For file-based distribution servers, this MAY for example be achieved by a symlink.
```
MAY is weak enough
```suggestion

  - For file-based distribution servers, this MAY be achieved by using a symlink.
    Redirects MUST NOT be used (cf. section [sec](#requirement-9-well-known-url-for-provider-metadata-json)).

oxisto replied:
done

oxisto commented:
> Please also add:
> 
> * [ ]  Transition announcement as subsubsection; the process as sibling subsubsection
> * [ ]  SHOULD announce transition period with 3 dates:
>   
>   * [ ]  `maintained_from` 2.1
>   * [ ]  roll-over date 2.0 - 2.1
>   * [ ]  `maintained_until` 2.0
> * [ ]  MAY announce two optional dates:
>   
>   * [ ]  `available_from` 2.1
>   * [ ]  `available_until` 2.0
> * [ ]  MAY use symlink for CSAF 2.1 PMD to link as current PMD; MUST NOT use redirects as stated in requirement 7.1.9
> * [ ]  SHOULD convert all CSAF 2.0 to CSAF 2.1 data (at least automatically)
> * [ ]  as optional step: move CSAF 2.0 files as archive to `.well-known/csaf/archive/v2.0.tar.bz2`

I think I have addressed all of the issues

oxisto commented on csaf_2.1/json_schema/aggregator_json_schema.json:
```diff
@@ -225,6 +198,18 @@
       "type": "string",
       "format": "date-time"
+    },
+    "maintained_from": {
```
I don't really like the wording on this :(

tschmidtb51 replied:
Me neither :-(

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,51 @@
+  - The roll-over-date at which CSAF 2.1 comes authoritative but CSAF 2.0 is still supported.
+  - The end of the transition period, after which CSAF 2.0 is not supported any more.
+
+### Process of transitioning `provider-metadata.json`
```
```suggestion
### Transition Process for a CSAF provider

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,51 @@
+- At the end of the transition period, it is RECOMMENDED to archive the remaining CSAF 2.0 documents, for example at
+  `.well-known/csaf/archive/v2.0.tar.bz2`
+
+### Process of transitioning `aggregator.json`
```
```suggestion
### Transition Process for a CSAF Aggregator

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,51 @@
+  - The URL SHALL also be added to the security.txt, if used.
+- At the begin of the transition period, a `provider-metadata.json` in conformance to CSAF 2.0 SHOULD be placed at `/.well-known/csaf/provider-metadata.json`.
+  - The content of the resource SHALL be equal to the resource accessible at `/.well-known/csaf/v2.0/provider-metadata.json`.
+  - For file-based distribution servers, this MAY be achieved by using a symlink.
```
```suggestion
  - For file-based distribution servers, this MAY be achieved by using a symlink.
    Redirects SHALL NOT be used (cf. to requirement [sec](requirement-9-well-known-url-for-provider-metadata-json))

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,51 @@
+- Sometime before the roll-over-date, all existing CSAF 2.0 documents SHOULD be converted to CSAF 2.1.
+- A the roll-over-date, a `provider-metadata.json` in conformance to CSAF 2.1 SHOULD be placed at `/.well-known/csaf/provider-metadata.json`.
+  - The content of the resource SHALL be equal to the resource accessible at `/.well-known/csaf/v2.1/provider-metadata.json`.
+  - For file-based distribution servers, this MAY be achieved by using a symlink.
```
```suggestion
  - For file-based distribution servers, this MAY be achieved by using a symlink.
    Redirects SHALL NOT be used (cf. to requirement [sec](requirement-9-well-known-url-for-provider-metadata-json))

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,51 @@
+    being authoritative.
+  - The roll-over-date at which CSAF 2.1 comes authoritative but CSAF 2.0 is still supported.
+  - The end of the transition period, after which CSAF 2.0 is not supported any more.
+
```
```suggestion

The announcement MAY contain also the following information:

- The first date, when CSAF 2.1 documents are available and automatically retrievable from the server through a CSAF 2.1 `provider-metadata.json`.
- The last date, when CSAF 2.0 documents are available and automatically retrievable from the server through a CSAF 2.0 `provider-metadata.json`.

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -11,7 +11,13 @@ Different scenarios can be encountered:
+The following process SHOULD be followed within the transition period:
 - A `provider-metadata.json` in conformance to CSAF 2.0 SHOULD be placed at `/.well-known/csaf/v2.0/provider-metadata.json`.
   - Its `canonical_url` MUST be set to an URL corresponding to the `/.well-known/csaf/v2.0/provider-metadata.json` path.
   - The property `maintained_until` SHOULD be set to the end of the transition period.
```
Yes. So it is even more important to introduce it right now....

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,74 @@
+
+The following rules apply for the archival of CSAF document from a previous version:
+
+- This archive SHOULD be located in `/.well-known/csaf/archive/` and use the file name `v2.0.zst`, `v2.0.tar.bz2` or `v2.0.tar.xz`.
```
My bad, zst is just the compression - the tar is needed...
```suggestion
- This archive SHOULD be located in `/.well-known/csaf/archive/` and use the file name `v2.0.tar.zst`, `v2.0.tar.bz2` or `v2.0.tar.xz`.

sthagen reviewed (approved):
Let's shipt it.

tschmidtb51 commented:
@sthagen Do we need to formally vote on this one?

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,74 @@
+Different scenarios can be encountered:
+
+- Providers will continue to use CSAF 2.0.
+  This SHOULD be avoided, as the CSAF ecosystem greatly profits from adoption to the new standard version.
```
Shouldn't that be "adoption of" instead of "adoption to"?
```suggestion
  This SHOULD be avoided, as the CSAF ecosystem greatly profits from adoption of the new standard version.

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,74 @@
+
+### Announcing the Transition
+
+In the last scenario, a temporary parallel distribution of CSAF 2.0 and CSAF 2.1 documents and provider metadata is recommended.
```
Let's make it normative.
```suggestion
In the last scenario, a temporary parallel distribution of CSAF 2.0 and CSAF 2.1 documents and provider metadata is RECOMMENDED.

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,74 @@
+
+- The begin of the transition period, where the provider is starting to serve CSAF 2.1 documents, while CSAF 2.0 being authoritative.
+- The roll-over-date at which CSAF 2.1 becomes authoritative but CSAF 2.0 is still supported.
+- The end of the transition period, after which CSAF 2.0 is not supported any more.
```
```suggestion
- The end of the transition period, after which CSAF 2.0 is not supported, i.e. maintained, any more.

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,74 @@
+
+The announcement MAY contain also the following information:
+
+- The first date, when CSAF 2.1 documents are available and automatically retrievable from the server through a CSAF 2.1 `provider-metadata.json`.
```
Let's provide some additional motivation for this.
```suggestion
- The first date, when CSAF 2.1 documents are available and automatically retrievable from the server through a CSAF 2.1 `provider-metadata.json`.
  > This might be a public beta prior to the begin of the transition period.
  > However, those documents should not be used in production as they may be unstable due to their beta status.

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,74 @@
+In the last scenario, a temporary parallel distribution of CSAF 2.0 and CSAF 2.1 documents and provider metadata is recommended.
+The provider SHOULD announce a transition period containing three points in time:
+
+- The begin of the transition period, where the provider is starting to serve CSAF 2.1 documents, while CSAF 2.0 being authoritative.
```
Let's clarify that against the one below...
```suggestion
- The begin of the transition period, where the provider is starting to serve CSAF 2.1 documents, while CSAF 2.0 being authoritative.
  > It is expected that the CSAF 2.1 files can be used in production from this point onward.

tschmidtb51 commented on csaf_2.1/prose/edit/src/distributing-04-transition-20-21.md:
```diff
@@ -0,0 +1,74 @@
+The announcement MAY contain also the following information:
+
+- The first date, when CSAF 2.1 documents are available and automatically retrievable from the server through a CSAF 2.1 `provider-metadata.json`.
+- The last date, when CSAF 2.0 documents are available and automatically retrievable from the server through a CSAF 2.0 `provider-metadata.json`.
```
Let's clarify that as well.
```suggestion
- The last date, when CSAF 2.0 documents are available and automatically retrievable from the server through a CSAF 2.0 `provider-metadata.json`.
  > This date is usually at the end or after the transition period.

oxisto replied:
I thought this markes the end of the transition period as we have defined it now. I would not include the "after the" part.

tschmidtb51 reviewed (changes requested):
@oxisto Sorry - it is a maturing process.

Please also include the DNS path as discussed

tschmidtb51 commented:
Introduced in the mailing list at: https://groups.oasis-open.org/discussion/transition-from-csaf-20-to-csaf-21-991

tschmidtb51 reviewed (commented):
Some minor changes - other than that, I think this is TC-ready.

tschmidtb51 commented:
@oxisto Thank you for your contribution.

justmurphy commented on csaf_2.1/prose/edit/src/distributing-01-requirements.md:
```diff
@@ -141,6 +150,8 @@ CSAF: https://domain.tld/security/csaf/provider-metadata.json
 It is possible to advertise more than one `provider-metadata.json` by adding multiple `CSAF` fields,
 e.g. in case of changes to the organizational structure through merges or acquisitions.
```
small nit 

```suggestion
e.g. in case of changes to the organizational structure through mergers and acquisitions.
```
